torch/csrc/acc/Module.cpp

```diff
@@ -0,0 +1,205 @@
+#include <torch/csrc/acc/Module.h>
+
+#include <ATen/ATen.h>
+#include <torch/extension.h>
+
+#include <ATen/detail/PrivateUse1HooksInterface.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <torch/csrc/utils/pybind.h>
+
+namespace py = pybind11;
+
+namespace torch::acc {
+
+// python hook interface
+struct PythonHooks final : public at::PrivateUse1HooksInterface {
+  using at::PrivateUse1HooksInterface::PrivateUse1HooksInterface;
+  bool hasPrimaryContext(c10::DeviceIndex device_index) const override {
+    PYBIND11_OVERRIDE_PURE_NAME(
+        bool,
+        at::PrivateUse1HooksInterface,
+        "has_primary_context",
+        hasPrimaryContext,
+        device_index);
+  }
+
+  bool isBuilt() const override {
+    PYBIND11_OVERRIDE_PURE_NAME(
+        bool, at::PrivateUse1HooksInterface, "is_built", isBuilt, );
+  }
+
+  bool isAvailable() const override {
+    PYBIND11_OVERRIDE_PURE_NAME(
+        bool, at::PrivateUse1HooksInterface, "is_available", isBuilt, );
+  }
+
+  // TODO(qihqi): these is not supported from python yet
+  const at::Generator& getDefaultGenerator(
+      c10::DeviceIndex device_index) const override {
+    return at::PrivateUse1HooksInterface::getDefaultGenerator(device_index);
+  }
+
+  at::Generator getNewGenerator(
+      c10::DeviceIndex device_index = -1) const override {
+    return at::PrivateUse1HooksInterface::getNewGenerator(device_index);
+  }
+
+  at::Device getDeviceFromPtr(void* data) const override {
+    return at::PrivateUse1HooksInterface::getDeviceFromPtr(data);
+  }
+
+  bool isPinnedPtr(const void* data) const override {
+    return at::PrivateUse1HooksInterface::isPinnedPtr(data);
+  }
+
+  at::Allocator* getPinnedMemoryAllocator() const override {
+    return at::PrivateUse1HooksInterface::getPinnedMemoryAllocator();
+  }
+};
+
+struct PythonDeviceGuard final : public c10::impl::DeviceGuardImplInterface {
```

Comment:
I guess you're ok with having only a single device and nothing being indirected to python for now here?